protocol/daemons/server/types/health_checker.go

```diff
@@ -0,0 +1,169 @@
+package types
+
+import (
+	"github.com/cometbft/cometbft/libs/log"
+	"github.com/dydxprotocol/v4-chain/protocol/daemons/types"
+	libtime "github.com/dydxprotocol/v4-chain/protocol/lib/time"
+	"sync"
+	"time"
+)
+
+// timestampWithError couples a timestamp and error to make it easier to update them in tandem. The
+// timestampWithError will track the timestamp of the first error in a streak of errors, but keeps a record of the
+// most recent error. This is useful for determining how long a service has been unhealthy, as well as the current
+// state of the service.
+type timestampWithError struct {
+	timestamp time.Time
+	err       error
+}
+
+func (u *timestampWithError) Update(timestamp time.Time, err error) {
+	// If the timestamp is zero, this is the first update, so set the timestamp.
+	if u.timestamp.IsZero() {
+		u.timestamp = timestamp
+	}
+	u.err = err
+}
+
+func (u *timestampWithError) Reset() {
+	u.timestamp = time.Time{}
+	u.err = nil
+}
+
+func (u *timestampWithError) IsZero() bool {
+	return u.timestamp.IsZero() && u.err == nil
+}
+
+func (u *timestampWithError) Timestamp() time.Time {
+	return u.timestamp
+}
+
+func (u *timestampWithError) Error() error {
+	return u.err
+}
+
+// healthCheckerMutableState tracks the current health state of the HealthCheckable, encapsulating all mutable state
+// into a single struct for ease of synchronization.
+type healthCheckerMutableState struct {
+	// lock is used to synchronize access to mutable state fields.
+	lock sync.Mutex
+
+	// mostRecentSuccess is the timestamp of the most recent successful health check.
+	// Access to mostRecentSuccess is synchronized.
+	mostRecentSuccess time.Time
+
+	// mostRecentFailureStreakError tracks the timestamp of the first error in the most recent streak of errors, as well
+	// as the most recent error. It is updated on every error and reset every time the service sees a healthy response.
+	// This field is used to determine how long the daemon has been unhealthy. If this timestamp is nil, then either
+	// the service has never been unhealthy, or the most recent error streak ended before it could trigger a callback.
+	// Access to mostRecentFailureStreakError is synchronized.
+	mostRecentFailureStreakError timestampWithError
```

Comment:
```suggestion
	lastErrWithFirstErrTimestamp timestampWithError
```

Reply:
I changed the type to `errorStreak` and call this `mostRecentErrorStreak`.